lib/SIL/SILConstants.cpp

```diff
@@ -354,6 +377,56 @@ UnknownReason SymbolicValue::getUnknownReason() const {
   return value.unknown->reason;
 }
 
+//===----------------------------------------------------------------------===//
+// Enums
+//===----------------------------------------------------------------------===//
+
+namespace swift {
+
+/// This is the representation of a constant enum value with payload.
+struct EnumWithPayloadSymbolicValue final {
+  /// The enum case.
+  EnumElementDecl *enumDecl;
+  SymbolicValue payload;
+
+  EnumWithPayloadSymbolicValue(EnumElementDecl *decl, SymbolicValue payload)
+      : enumDecl(decl), payload(payload) {}
+
+private:
+  EnumWithPayloadSymbolicValue() = delete;
+  EnumWithPayloadSymbolicValue(const EnumWithPayloadSymbolicValue &) = delete;
+};
+} // end namespace swift
+
+/// This returns a constant Symbolic value for the enum case in `decl` with a
+/// payload.
+SymbolicValue
+SymbolicValue::getEnumWithPayload(EnumElementDecl *decl, SymbolicValue payload,
+                                  ASTContext &astContext) {
+  assert(decl && payload.isConstant());
+  auto rawMem = astContext.Allocate(sizeof(EnumWithPayloadSymbolicValue),
+                                    alignof(EnumWithPayloadSymbolicValue));
```

Comment:
@marcrasi  I have a more general question. It would be helpful if you can throw some light on why we have to allocate the symbolic values in the allocator of the `astContext`, and not bind its lifetime to the interpreter instance i.e, ConstExprEvaluator instance. (I think you clarified this in the earlier PRs, but it slipped my mind.) This essentially means these objects are never cleaned up (or at least retained for a long time). On the upside,  I see that this enables querying the symbolic state of a SILValue long after interpreter has died. I wonder whether there is an inherent reason why this is needed or is this a design choice motivated by other use cases (possibly Swift for Tensorflow usecases).

Reply:
One of our passes does store `SymbolicValue`s into a SILInstruction that lives longer than the ConstExprEvaluator. (see `GraphOperationInst` in SILInstruction.h in the tensorflow branch if you're intersted, and look at the call to `constantEvaluator.computeConstantValues` in TFDeabstraction.cpp).

It should be pretty easy to clone all the memory into a longer-lived allocator before we do that though. It would be good to write APIs such that it's hard to forget to do this. Maybe all the public ConstExprEvaluator functions could clone all the memory for the returned SymbolicValues into the astcontext by default. There could be an argument allowing callers who don't actually need long-lived SymbolicValues to opt out of the cloning.

Reply:
An alternative is that we can also provide a suitable allocator as a parameter to the constructor of ConstExprEvaluator (e.g. as Lambda or something else). But, this is something we can discuss in a separate PR.

Reply:
@marcrasi I am thinking of moving some design discussions such as the above to Swift forum (compiler development section), so that we can discuss these things is more public visibility
and hopefully involve more people. :-)